py/sentry_streams/pipeline.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+from sentry_streams.adapters.stream_adapter import StreamAdapter
+
+
+class RuntimeTranslator:
+    def __init__(self, runtime_adapter: StreamAdapter):
+        self.adapter = runtime_adapter
+
+    def translate_source(self, step_config: Mapping[str, Any]) -> Any:
+        translated_fn = getattr(self.adapter, "source")
+
+        return translated_fn(step_config)
+
+    def translate_with_input(
+        self, step: WithInput, step_config: Mapping[str, Any], stream: Any
+    ) -> Any:
+        next_step = step.step_type
+        translated_fn = getattr(self.adapter, next_step)
+
+        return translated_fn(step_config, stream)
+
+
+class PipelineGraph:
+    def __init__(self) -> None:
+        self.steps: dict[str, Step] = {}
+        self.incoming_edges: dict[str, list[str]] = defaultdict(list)
+        self.outgoing_edges: dict[str, list[str]] = defaultdict(list)
+        self.sources: list[Source] = []
+
+    def register(self, step: Step) -> None:
+        assert step.name not in self.steps
+        self.steps[step.name] = step
+
+    def register_edge(self, _from: Step, _to: Step) -> None:
```

Comment:
nit: why were `from` and `to` called `_from` and `_to` ?

Reply:
Not sure but can change

Reply:
I think it was written that way because `from` is reserved keyword

Reply:
oh right